File: answers/other/answers/lintcode/Valid Sudoku.cpp
class Solution {
public:
    /**
      * @param board: the board
      * @return: wether the Sudoku is valid
      */
    bool isValidSudoku(const vector<vector<char>>& board) {
		if (0 == board.size() || 0 == board[0].size()) {
			return true;
		}
        for (size_t i = 0; i < board.size(); ++ i) {
			set<int> sudoku;
			for (size_t j = 0; j < board[i].size(); ++ j) {
				if ('.' != board[i][j] && sudoku.end() != sudoku.find(board[i][j])) {
					return false;
				}
				sudoku.insert(board[i][j]);
			}
		}
        for (size_t i = 0; i < board[0].size(); ++ i) {
			set<int> sudoku;
			for (size_t j = 0; j < board.size(); ++ j) {
				if ('.' != board[j][i] && sudoku.end() != sudoku.find(board[j][i])) {
					return false;
				}
				sudoku.insert(board[j][i]);
			}
		}
		for (size_t i = 0; i < board.size(); i += 3) {
			for (size_t j = 0; j < board[i].size(); j += 3) {
				set<int> sudoku;
				for (size_t m = 0; m < board.size() / 3 && i + m < board.size(); ++ m) {
					for (size_t n = 0; n < board[i + m].size() / 3 && j + n < board[i + m].size(); ++ n) {
						if ('.' != board[i + m][j + n] && sudoku.end() != sudoku.find(board[i + m][j + n])) {
							return false;
						}
						sudoku.insert(board[i + m][j + n]);
					}
				}
			}
		}
		return true;
    }
};

File: answers/leetcode/Valid Sudoku/Valid Sudoku.java
import java.util.Set;

class Solution {
    public boolean isValidSudoku(char[][] board) {
        List<Integer> sudoku;
        for (int i = 0; i < 9; i++) {
            sudoku = new LinkedList<>();
            for (int j = 0; j < 9; j++) {
                if (Character.isDigit(board[i][j])) {
                    sudoku.add(Character.getNumericValue(board[i][j]));
                }
            }
            if (!this.check(sudoku)) {
                return false;
            }

            sudoku = new LinkedList<>();
            for (int j = 0; j < 9; j++) {
                if (Character.isDigit(board[j][i])) {
                    sudoku.add(Character.getNumericValue(board[j][i]));
                }
            }
            if (!this.check(sudoku)) {
                return false;
            }

            if (i % 3 == 0) {
                for (int j = 0; j < 9; j += 3) {
                    sudoku = new LinkedList<>();
                    for (int k = 0; k < 3; k++) {
                        for (int m = 0; m < 3; m++) {
                            if (Character.isDigit(board[i + k][j + m])) {
                                sudoku.add(Character.getNumericValue(board[i + k][j + m]));
                            }
                        }
                    }
                    if (!this.check(sudoku)) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    private boolean check(List<Integer> sudoku) {
        Set<Integer> set = new HashSet<Integer>(sudoku);
        return set.size() == sudoku.size();
    }
}